### ISSUE TITLE
Performance issue ?

### ISSUE THREAD

**Issue by Narsil** [rustformers/llama-rs#97]
Hello, thanks for the great work.

I'm trying your fork, and it seems to be much slower than the original llama.cpp on M1.

Anything known or that I'm forgetting during compilation ?:

```
RUSTFLAGS="-C target-cpu=native" cargo run --release -- -m ~/ggml-model-q4_0.bin.back -p "Hello" -n 10
```

That should be enough, no ?

**Comment by philpax** [comment]
I've fixed this in #81 - it's a combination of two issues (not linking with Accelerate) and `num_cpu::get_physical` including the efficiency cores, which massively slows things down. Try `-t 8` (or equivalent for your CPU) to see much better performance before the PR.

**Comment by saites** [comment]
@Narsil, There was, at one time, [an issue](https://github.com/rust-lang/rust/issues/93889) with `rustc`'s feature detection for the M1. That issue was closed awhile ago, so it probably isn't a problem anymore, but it might be worth checking whether you get better performance from specifying it explicitly:

```
RUSTFLAGS="-C target-cpu=apple-m1" cargo build --release
```

[The Perf Book](https://nnethercote.github.io/perf-book/build-configuration.html#using-cpu-specific-instructions) suggests comparing the output of these two commands as well:

```
> rustc --print cfg
> rustc --print cfg -C target-cpu=native
```

**Comment by philpax** [comment]
That won't have any impact - all of the computation is in `ggml`, which is C.

**Comment by saites** [comment]
Ah, fair enough. I did some profiling at one point, and unsurprisingly, most of the time is spent in the `ggml` calls -- particularly matrix multiplication. I was originally looking into optimizations around sampling. While I think there's some room for improvement, particularly with memory use, it's just dwarfed by the actual inference runtime. Again, not exactly a surprise. 

I will say I did get a bit better performance after profile-guided optimization (using `llvm-profdata`). That was with an older version of llama -- and I've got an i7-9700KF, so definitely not sure to what extent that would translate.
